### ISSUE TITLE
Feature Suggestion: Multi-region support for ELS

### ISSUE THREAD

**Issue by MattJeanes** [featbit/featbit#665]
**Is your feature request related to a problem? Please describe.**

We are trying to scale out FeatBit across multiple regions, in particular the evaluation server (ELS) but have found it is dependent on the same Redis instance that the API server uses, rather than being capable of using different ones.

**Describe the solution you'd like**

We would like to be able to use a separate Redis instance for the ELS so apart from the first call / startup of the ELS it should be essentially self-contained and not require talking to the main API / mongo / redis instance.

**Describe alternatives you've considered**

We have also considered using the FeatBit Agent but the manual sync requirement is a dealbreaker for us, and it also seems somewhat unfinished with a lack of a dockerfile/container/Helm chart.

Looking through the code of the ELS we can see it uses a hybrid store which prefers Redis but can fall back to MongoDB if Redis is unavailable for some reason. This is closer to what we're looking for, but 1. it really doesn't like you removing the Redis connection string right now (we're using Pro model so Kafka for messaging) and 2. it doesn't cache the data it loads from MongoDB into memory or Redis.

We have also considered, and are still considering writing our own program to make this work, but ideally we want to work with you on a good solution that benefits everyone.

**Additional context**

Let me describe our rough infrastructure setup so you can see what we're talking about:

So we have a "global" region for resources which must sit in one location, and this is where FeatBit is sat currently. Then we have a bunch of other regions around the global, for example EU or US that contain apps that need to read feature flags using the FeatBit client. We want to reduce latency and improve resiliency against our "global" region potentially going down by not having anything in our other various regions directly relying on it if possible.

This is what we were hoping the evaluation server could do, act as essentially a middle-man between our apps and the "global" featbit API / mongo / etc and use it's own Redis instance to cache data locally to that particular region. What we found though is that the featbit API updates Redis directly so it doesn't get updated by the ELS when it receives the update message over Kafka, that appears to only be for updating live connected clients as far as we can tell.

As mentioned we explored the FeatBit agent which seems to be closer to the design we want, but seems unfinished and requires manual syncing at this time. We're happy for the individual instances of the ELS in each region to reach out to the API server or mongo etc to grab the initial feature flag data, but from then on it should be able to serve them from it's own Redis cache without touching that "global" region. And of course any updates should come through the Kafka queue and also get saved in it's own Redis cache too.

**Is this feature something you're interested in working on?**

Yes, we are interested / happy to work with you on implementing this if we can agree on a design that everyone is happy with, so this issue is me opening the conversation to see what you think about this.

If we can't agree on a design then we will likely build our own app for this use case, but hopefully as mentioned we can contribute these improvements back for everyone's benefit 🙂 

Also this is all based off a light reading of the FeatBit source code so if we have any assumptions wrong here please correct us!

Thank you!

**Comment by MattJeanes** [comment]
Hey @deleteLater / team!

That's a great point about the multiple ELS instances in one region. It would break the services design of being scalable really doing that, so we probably shouldn't go for that option.

As for upgrading the FeatBit Agent, we think this is indeed better but there is some additional concerns I wanted to run by you:

1. The FeatBit agent currently only caches to a sqlite database, which in a containerised world might mean it gets wiped every time  the container is re-created. It would be fantastic if we could make this sync to a Redis instance in the same way that the API does, which leads on to...
2. It seems that the FeatBit agent isn't really designed to be horizontally scaled which presents a challenge compared to using the ELS as we would like to have at minimum 2 containers for reliability, and for scale as we'll be connecting these to potentially hundreds of containers using feature flags.
3. The FeatBit agent does not currently appear to support feeding back analytics to the data analytics stuff e.g. clickhouse

Something that could be quite interesting to solve these issues is if the FeatBit agent cached to Redis in the same format that the API does then we could use the ELS as we originally planned. That would allow us to easily horizontally scale, fix feeding back analytics and basically solve all the problems as best as I can tell.

So in effect, the FeatBit Agent essentially becomes a broker to sync the flags into separate Redis instances, which the ELS can then use. And if Redis is unavailable for some reason the ELS can fall back to the central MongoDB store as it currently would.

Also, I'm a DevOps / Platform Engineer so I can definitely get the containerisation / Helm chart side of things done for you. We also have another backend / C# dev who I might be able to bring on to help out with the actual C# code side of things.

How do you feel about this?

**Comment by deleteLater** [comment]
Sorry for the late response. We haven't had much time to work on this project lately.

I think using the FeatBit Agent as "a broker to sync the flags into separate Redis instance" is a good solution.

To implement that we need to add a new working mode called "broker" to FeatBit Agent. Currently we've defined [two working modes](https://docs.featbit.co/relay-proxy/featbit-agent#concepts): Agent[not implemented yet] and Offline. When working in "broker" mode, the FeatBit Agent will pull data updates from the API server and update the data in Redis, which ELS can then use.

Do you think that's ok?

**Comment by MattJeanes** [comment]
Hey, no worries about the late response although we have started to work on our own solution for this in the meantime.

Our own solution we're exploring at the moment is pretty similar to how the agent is described in the docs and uses a single connection to the ELS to pull all the latest flag information which it'll also cache in its own Redis store and then downstream clients can connect to it in place of the ELS to provide that resiliency and improved latency.

We're not done yet, and as such that design is not final, but if this works we're happy to contribute this back to be integrated into the proper FeatBit Agent if that's something you're interested in.

Note that currently it's a standalone .NET 8 agent app we've built rather than modifying the existing agent for now, but they could be merged together in the future.

The difference between our and your ideas seems to be that your version still uses the ELS downstream while ours uses it upstream and takes it's place for end clients.

One problem your version does solve though is that scaling out ours will also scale out saving the data to Redis multiple times which isn't a big deal but maybe a little hacky, and it's from what I understand the reason that the API service saves to Redis instead of the ELS.

We wanted to write a solution that worked without needing to modify any FeatBit code which is why we are using the ELS upstream to grab all the flag data from but your idea may work better as a proper long term solution.

Ultimately the problem we're dealing with is that the API only updates the one Redis instance when flags are changed which was the main reason we couldn't just drop the ELS into multiple regions with their own Redis connection strings.

We do note that the ELS currently has a mongo connection string, but if I remember correctly from reading the code it's only used as a backup if it can't retrieve the flags from Redis for some reason. This is fine but we need to make sure this is optional and the app can start up and work with only a Redis connection string set.

In other words we want to make sure that even if our region containing the primary FeatBit infrastructure goes offline the individual regions used by our product can at least still start up and serve the latest feature flags from their local Redis caches.

I think the main issue we have with the current FeatBit Agent is that the storing/loading from Redis is not yet implemented - only in memory - so if the agent has to restart it won't work if it can't reach the API server, and that the sync is manual and not real time. If we could fix those two things I wonder what the difference would be between the broker and agent mode, so I'm curious if it would be perhaps better to solve those problems instead of a new mode?

**Comment by deleteLater** [comment]
> it's from what I understand the reason that the API service saves to Redis instead of the ELS

I don't remember why, but it was certainly one of the reasons we did it.

> it's only used as a backup if it can't retrieve the flags from Redis for some reason. This is fine but we need to make sure this is optional and the app can start up and work with only a Redis connection string set.

Yes, mongodb is only used as a fallback but currently the mongodb connection string must be set. To make it optional, we need to 

1. update `HealthCheckBuilderExtensions.cs`  
2. we should make the `MongoDbClient` optional or initialize it lazily

> If we could fix those two things I wonder what the difference would be between the broker and agent mode, so I'm curious if it would be perhaps better to solve those problems instead of a new mode?

Yeah, I was mistaken that. We just need to implement agent mode.

Currently implemented **offline mode**

1. Use sqlite as persistent store 
2. The API server can "bootstrap" the agent, passing feature flags and segments data in and the agent stores them in both memory and sqlite, so the next time when the agent starts, it populates the memory cache by reading the sqlite.
3. Store insight data (via '/track' api) and `SyncHistory` in sqlite

If we want to finish the **agent mode**, I think there's a lot of work needs to be done. 

How about we just write a `data-synchronizer` program to synchronize data from the global MongoDb to regional Redis?

**Comment by ghelyar** [comment]
We were trying to use the evaluation server as an agent because the agent does not currently have the agent mode implemented. and if we could run the evaluation server in that way then we would not need a separate agent running in agent mode.

As @MattJeanes said, we're currently implementing our own agent that runs in agent mode so we would be happy to contribute that back into the FeatBit agent, if agent mode is still needed.

It really depends on how much effort it would be to write a data synchroniser vs agent mode, and whether it would be better long term to run the evaluation server in this way or have a separate agent with agent mode.

**Comment by deleteLater** [comment]
Update on this: We can utilize [Webhooks](https://docs.featbit.co/integrations/webhooks) to "sync" data updates to agent, instead of polling data updates from the server. If we do so, we need to bootstrap the agent on startup.

**Comment by ghelyar** [comment]
Right now this is what we have implemented:

FeatBit is self-hosted in a central region, and multiple other regions connect to it via an agent that we have written, which has a subset of the functionality of the described [FeatBit agent mode](https://docs.featbit.co/relay-proxy/featbit-agent#agent-mode), which does not exist yet in the FeatBit agent. (IMHO the documentation for the agent should really say that agent mode is not implemented yet, because right now it recommends using a mode that does not exist.)

Many replicas of many services connect to the agent, and the agent keeps a single WebSocket connection to the FeatBit Evaluation Server for each environment secret, as long as there is at least one client for that environment. When there is a change to the flags, the evaluation server sends a message to the agents, and the agents send messages to the clients. When there area no more clients for that environment, the agent disconnects from the evaluation server, and when a new client connects the agent makes a new connection to the evaluation server. In the case of an error the client reconnects to the agent and the agent reconnects to the server.

A redis cache is used to keep the last known state, but is only read as a fallback if no connection to the evaluation server can be made, so that new replicas of the agent can be started even if the evaluation server is not available. A memory cache also sits in front of the redis cache, so that clients get a quick response as soon as they connect.

Here's how the services are deployed across the regions
![Image](https://github.com/user-attachments/assets/9d5897f1-94c9-436b-96a9-df628fb5fc44)

Here's how the agent is implemented
![Image](https://github.com/user-attachments/assets/5f8aa940-9b6b-4b83-8837-e37a9f6444f4)

We have not currently implemented the events or the client SDK because we are not using them yet, but when we start using events we expect to have to push events directly to Kafka from the regional agents, as the number of events would likely overwhelm the evaluation server in the central region.

The other thing described in the agent mode which this does not do yet is data protection/anonymisation, because it does not capture any events from the client yet anyway.

We're happy to contribute what we've done so far back to the FeatBit agent if you want it.

**Comment by cosmic-flood** [comment]
Hi @ghelyar 

I think you're right—we need to update some parts of the FeatBit Agent documentation. Here's what the FeatBit Agent currently supports:

1. Both client-side and server-side SDKs can connect to the FeatBit Agent in the same way they would connect to the evaluation server.
2. We have to register the Agent with the FeatBit service and manually synchronize feature flag configurations between the remote server and the Agent.
3. Each Agent instance uses its own SQLite file so it can continue to function after a restart or if it cannot connect to the remote server.

We initially designed this architecture because our early customers needed to host the full service in a network-isolated environment for their clients. As a result, they only synchronized the Agent manually when possible.

**Your implementation makes more sense for what the Agent is ultimately supposed to do**. Contributing what you’ve built would be super valuable.

I have a few questions that aren’t entirely clear to me:

1. Is there only a single Agent instance, or are there multiple instances? Does only one Agent connect to the FeatBit evaluation server while the replicas connect to this Agent?
2. I assume the Redis cache is shared among multiple Agent replicas. Does only one Agent write to Redis? If that Agent fails, what happens?

I think the FeatBit team could build upon your solution and combine it with our work to enhance client SDK connectivity and data protection. We can then focus on event transmission and any other features that are important to you, us, and developers worldwide.

@deleteLater @cosmos-explorer @dsun0720  what do you think about it?

**Comment by ghelyar** [comment]
> 1. Is there only a single Agent instance, or are there multiple instances? Does only one Agent connect to the FeatBit evaluation server while the replicas connect to this Agent?

We have multiple replicas of the agent in each region, which we can scale automatically, but this number is much smaller than the number of end services that connect. For example, 2 evaluation servers, 10 agents, 100 consuming services (not real numbers).

In theory agents could connect to other agents but we haven't needed to do that.

> 2. I assume the Redis cache is shared among multiple Agent replicas. Does only one Agent write to Redis? If that Agent fails, what happens?

Right now when the evaluation server publishes a change, all of the agents in the same region write to the same redis cache - there is some duplication in the writes here but it's a simple solution, there aren't that many agents, and in theory they should all have the same data anyway.

At the moment this cache is really only used as an emergency backup, with the evaluation server being the real source of truth. This is just so that if we have an outage in the central region, the other regions can continue to operate, including adding new replicas of the agent. If we didn't store the current state in each region then existing agents would continue to work but we would not be able to add more replicas.

It needs a bit more polish and time in production before it's ready to contribute back, as we have not migrated many services to FeatBit yet, but it's not that far off.

**Comment by deleteLater** [comment]
That's amazing! We really appreciate your work so far ❤ @ghelyar 

We can have more discussion about this one on a dedicated daily-meting. @cosmic-flood

**Comment by deleteLater** [comment]
related to #734

**Comment by deleteLater** [comment]
Hi @ghelyar, Auto agents is now available and you may give it a try.

ref: https://github.com/featbit/featbit/releases/tag/5.1.0
